From 6a1b5adf4da37f8d854935bc78d81545251f052f Mon Sep 17 00:00:00 2001
From: Rob Evers <revers@redhat.com>
Date: Mon, 13 Oct 2014 19:47:51 -0400
Subject: [scsi] lpfc: 10.4.8000.0: Fixed crash from page fault caused by use after rport delete

Message-id: <1413229672-8452-7-git-send-email-revers@redhat.com>
Patchwork-id: 97523
O-Subject: [RHEL7.1 e-stor PATCH 6/7] scsi: lpfc: 10.4.8000.0: Fixed crash from page fault caused by use after rport delete
Bugzilla: 1149846
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Ewan Milne <emilne@redhat.com>

From: Dick Kennedy <rkennedy@emulex.com>

Submitted by James Smart James.Smart@emulex.com

Fixed crash from page fault caused by use after rport delete

Last 2 hunks in this patch are from a different commit, and are already
in the linux-2.6 tree.

Signed-off-by: James Smart james.smart@emulex.com
Signed-off-by: Dick Kennedy dick.kennedy@emulex.com

http://marc.info/?l=linux-scsi&m=140976347215048&w=2
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/lpfc/lpfc_hbadisc.c b/drivers/scsi/lpfc/lpfc_hbadisc.c
index a43929f..2e113d4 100644
--- a/drivers/scsi/lpfc/lpfc_hbadisc.c
+++ b/drivers/scsi/lpfc/lpfc_hbadisc.c
@@ -150,6 +150,17 @@ lpfc_dev_loss_tmo_callbk(struct fc_rport *rport)
 
   /* If the WWPN of the rport and ndlp don't match, ignore it */
   if (rport->port_name != wwn_to_u64(ndlp->nlp_portname.u.wwn)) {
+   lpfc_printf_vlog(vport, KERN_ERR, LOG_NODE,
+    "6789 rport name %lx != node port name %lx",
+    (unsigned long)rport->port_name,
+    (unsigned long)wwn_to_u64(
+      ndlp->nlp_portname.u.wwn));
+   put_node = rdata->pnode != NULL;
+   put_rport = ndlp->rport != NULL;
+   rdata->pnode = NULL;
+   ndlp->rport = NULL;
+   if (put_node)
+    lpfc_nlp_put(ndlp);
    put_device(&rport->dev);
    return;
   }
@@ -285,6 +296,7 @@ lpfc_dev_loss_tmo_handler(struct lpfc_nodelist *ndlp)
  if (ndlp->nlp_sid != NLP_NO_SID) {
   warn_on = 1;
   /* flush the target */
+  ndlp->nlp_add_flag &= ~NLP_IN_DEV_LOSS;
   lpfc_sli_abort_iocb(vport, &phba->sli.ring[phba->sli.fcp_ring],
         ndlp->nlp_sid, 0, LPFC_CTX_TGT);
  }
-- 
1.7.1